server/src/sql-tools/schema-diff.ts

```diff
@@ -0,0 +1,449 @@
+}
+
+const setIsEqual = (source: Set<unknown>, target: Set<unknown>) =>
+  source.size === target.size && [...source].every((x) => target.has(x));
+
+const haveEqualColumns = (sourceColumns?: string[], targetColumns?: string[]) => {
+  return setIsEqual(new Set(sourceColumns ?? []), new Set(targetColumns ?? []));
+};
+
+const isSynchronizeDisabled = (source?: { synchronize?: boolean }, target?: { synchronize?: boolean }) => {
+  return source?.synchronize === false || target?.synchronize === false;
+};
+
+const withTypeCast = (value: string, type: string) => {
+  if (!value.startsWith(`'`)) {
+    value = `'${value}'`;
+  }
+  return `${value}::${type}`;
+};
+
+const isDefaultEqual = (source: DatabaseColumn, target: DatabaseColumn) => {
+  if (source.default === target.default) {
+    return true;
+  }
+
+  if (source.default === undefined || target.default === undefined) {
+    return false;
+  }
+
+  if (
+    withTypeCast(source.default, getColumnType(source)) === target.default ||
+    source.default === withTypeCast(target.default, getColumnType(target))
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
+/**
+ * Compute the difference between two database schemas
+ */
+export const schemaDiff = (
+  source: DatabaseSchema,
+  target: DatabaseSchema,
+  options: { ignoreExtraTables?: boolean } = {},
+) => {
+  const items = diffTables(source.tables, target.tables, {
+    ignoreExtraTables: options.ignoreExtraTables ?? true,
+  });
+
+  return {
+    items,
+    asSql: (options?: SchemaDiffToSqlOptions) => schemaDiffToSql(items, options),
+  };
+};
+
+export const diffTables = (
+  sources: DatabaseTable[],
+  targets: DatabaseTable[],
+  options: { ignoreExtraTables: boolean },
+) => {
+  const items: SchemaDiff[] = [];
+  const sourceMap = Object.fromEntries(sources.map((table) => [table.name, table]));
+  const targetMap = Object.fromEntries(targets.map((table) => [table.name, table]));
+  const keys = new Set([...Object.keys(sourceMap), ...Object.keys(targetMap)]);
+
+  for (const key of keys) {
+    if (options.ignoreExtraTables && !sourceMap[key]) {
+      continue;
+    }
+    items.push(...diffTable(sourceMap[key], targetMap[key]));
+  }
+
+  return items;
+};
+
+const diffTable = (source?: DatabaseTable, target?: DatabaseTable): SchemaDiff[] => {
+  if (isSynchronizeDisabled(source, target)) {
+    return [];
+  }
+
+  if (source && !target) {
+    return [
+      {
+        type: 'table.create',
+        tableName: source.name,
+        columns: Object.values(source.columns),
+        reason: Reason.MissingInTarget,
+      },
+      ...diffIndexes(source.indexes, []),
+      // TODO merge constraints into table create record when possible
+      ...diffConstraints(source.constraints, []),
+    ];
+  }
+
+  if (!source && target) {
+    return [
+      {
+        type: 'table.drop',
+        tableName: target.name,
+        reason: Reason.MissingInSource,
+      },
+    ];
+  }
+
+  if (!source || !target) {
```

Comment:
I'd change this to `!source && !target` as it'll then be robust to (accidentally) being moved around.

Reply:
That does something different. I need typescript to know that both are defined after this point.